onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.cc

```diff
@@ -0,0 +1,296 @@
+static Tensor DiagonalInnermostDims(const Tensor& input, bool preserve_innermost_dim_val, AllocatorPtr allocator) {
+  const char* input_data = reinterpret_cast<const char*>(input.DataRaw());
+  const auto& input_dims = input.Shape().GetDims();
+  auto rank = input_dims.size();
+  const size_t element_size_in_bytes = input.DataType()->Size();
+
+  // This is an internal method and we already have finished all validations in the calling method.
+  // We proceed without duplicating all validations again here.
+
+  // We have a minimalistic check here to make sure the innermost dims have the same dim value
+  // as the calling method may have done a transpose before calling this method
+  ORT_ENFORCE(input_dims[rank - 2] == input_dims[rank - 1],
+              "The innermost dims should have the same dim value to parse the diagonal elements");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(rank);
+
+  int64_t num_iterations = 1;  // Flatten the outermost dims - this will be the number of iterations
+  for (size_t i = 0; i < rank - 2; ++i) {
+    auto input_dim_value = input_dims[i];
+    num_iterations *= input_dim_value;
+    output_dims.push_back(input_dim_value);
+  }
+
+  if (preserve_innermost_dim_val) {
+    output_dims.push_back(1);
+    output_dims.push_back(input_dims[rank - 1]);
+  } else {
+    output_dims.push_back(input_dims[rank - 1]);
+    output_dims.push_back(1);
+  }
+
+  int64_t inner_stride = input_dims[rank - 1];        // offset to move over the innermost dim
+  int64_t base_stride = inner_stride * inner_stride;  // offset to move over all the axes except the 2 innermost dims
+
+  // Pass in allocator as that will be used as an allocator deleter by the framework
+  // and it will de-allocate the memory for this intermediate tensor when it goes out of scope
+  Tensor output(input.DataType(), output_dims, allocator);
+  char* output_data = reinterpret_cast<char*>(output.MutableDataRaw());
+
+  int64_t output_iter = 0;
+  // TODO: Parallelize this operation
+  for (int64_t i = 0; i < num_iterations; ++i) {
+    auto base_offset = i * base_stride;
+    for (int64_t j = 0; j < inner_stride; ++j) {
+      memcpy(output_data + output_iter * element_size_in_bytes,
+             input_data + (base_offset + j * inner_stride + j) * element_size_in_bytes,
+             element_size_in_bytes);
+      output_iter++;
+    }
+  }
+
+  return output;
+}
+
+Tensor Diagonal(const Tensor& input, int64_t dim_1, int64_t dim_2, AllocatorPtr allocator) {
+  const auto& dims = input.Shape().GetDims();
+  auto rank = static_cast<int64_t>(dims.size());
+
+  ORT_ENFORCE(rank >= 2 && dim_1 >= 0 && dim_2 >= 0 && dim_1 < rank &&
+                  dim_2 < rank && dim_1 != dim_2 && dims[dim_1] == dims[dim_2],
+              "Cannot parse the diagonal elements along dims ", dim_1, " and ", dim_2, " for input shape ", input.Shape());
```

Comment:
Possibly not necessary to check the dims are non-negative and < rank (not sure how that would be possible given it's an internal function and I assume the input to einsum would have been validated before this is called). Makes it harder to understand the purpose of this check.

Reply:
Sure. Removed obvious checks.